src/rgw/rgw_common.h

```diff
@@ -1297,6 +1297,7 @@ struct req_state {
   /* aws4 auth support */
   bool aws4_auth_needs_complete;
   bool aws4_auth_streaming_mode;
+  bool aws4_auth_presigned_url_bugfix;
```

Comment:
No really, both booleans have a similar name cause they work on the presigned url bug but they need to hold two different boolean states working in tandem. They will only contain the same state/value in the authorize_v4's second run when the admin turned the 'rgw_s3_auth_aws4_presigned_url_bugfix' flag on and the presigned url case happens.

Let me share in detail. The first boolean 'rgw_s3_auth_aws4_presigned_url_bugfix' is the configurable option. It helps to decide if we have to compute the second signature in the presigned url case only.

The second boolean 'aws4_auth_presigned_url_bugfix' is needed to drive the logic in authorize_v4 and, at the same time, keeping the function body with minimal modifications. You will see the only modification with this approach in authorize_v4 is checking this var when the using_qs boolean is 'true' (parameters shipping in the query string as in the case of presigned urls).
 
The key part is the 'aws4_auth_presigned_url_bugfix' needs to be initialized to 'false' and it will have that value but if it is a presigned url case *and* the configurable 'rgw_aws4_auth_presigned_url_bugfix' is also 'true'. In other words, authorize_v4 always uses 'aws4_auth_presigned_url_bugfix' as 'false' in the first run (first signature computation) and then, if it is a presigned url and the admin wants runs the runtime bugfix, it will run a second signature computation with the 'aws4_auth_presigned_url_bugfix' boolean to true.

You can consider 'rgw_s3_auth_aws4_presigned_url_bugfix' as the policy and 'aws4_auth_presigned_url_bugfix' as the mechanism.